Set acl permissions on workdir root to ensure multi-user access when umask is `0027`

carlosrodfern opened the pull request:
In machines with umask set to `0027` it is necessary to setup an acl to overide it so the workdir root directory stays multi-user access. This is specially necessary on machines accessed with a non-root user, that are hardened to CIS Level 1 guidelines.

It is happening in version 1.29.0.

Fixes: #2496

Pull Request Checklist

* [x] implement the feature
* [x] write the documentation (docstrings)
* [x] extend the test coverage
* [x] update the specification (N/A)
* [x] adjust plugin docstring
* [x] modify the json schema (N/A)
* [x] mention the version
* [x] include a release note (the PR title)

carlosrodfern commented:
This MR is just a draft, to bring this conceptual fix to the referred issue. There is some cleanup to do. I'm OK with a complete change of approach or even if this is not worth the fix if there is a workaround. That being said hardened machines are so common that the fix may actually be worth doing.

Thoughts?

happz commented on examples/plugins/provision.py:
```diff
@@ -84,6 +84,7 @@ def go(self):
 
         self._guest = GuestExample(data, name=self.name, parent=self.step)
         self._guest.start()
+        self._guest.setup()
```
Every plugin already calls `start()`. Can `start()` call `setup()` directly? That would give us just one method to remember to call in a plugin, instead of two.

`setup()` follows `start()`, with the exception of the `local` and `connect` plugins, and I can imagine these two could add their custom "start nothing and call `start()`, like the parent would" implementations.

carlosrodfern replied:
@happz , I thought of having this operation done in `start` directly or indirectly with another method. However, I read the description of `start` and hesitated. It appears to me that adding this functionality, which gets inside the machine itself to do things, increases the semantic of the `start` function, perhaps in a surprising way for a developer.

`start` a `guest` seems to be that the end goal is to have a guest turned on and access given, but the internal modification in order to prepare doesn't come to mind. `setup` a `guest`, on the other hand, can imply an internal modification of the VM. Naming and semantics definition is one of the hardest things in computer science, lol.

Thoughts?

happz replied:
> There are 2 hard problems in computer science: cache invalidation, naming things, and off-by-1 errors.

Anyway. Good point. I still think not giving developers a chance to forget such an important step in the process is worth the effort :) I'll think about it, maybe we could extend the semantics of `Guest.start()`, maybe rename it, who knows.

carlosrodfern replied:
I'm looking more into this one.

The "start" in artemis and mrack means "create". In test cloud means the provisioning.  The "setup" step here is really a "post" start step common to all of them. So we get then a "prestart" that does the create, or provisioning, and a "poststart" that does what I have called "setup". Whoever comes later with a new implementation of "start" will need to remember to also run the "poststart" after the machine is created.

Another place where I could put this is in the `PreparePlugin`, which has a `go` method which is always called in the subclasses via `super`.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1268,6 +1276,25 @@ def is_ready(self) -> bool:
+            parent = cast(Provision, self.parent)
+            ssh_command: tmt.utils.Command = self._ssh_command()
+
+            remote_commands: ShellScript = ShellScript(f'sudo dnf install -y acl')
```
Instead of `dnf`, `guest.facts.package_manager` should be used.

(We really, really should have "install me this" helpers, because tmt installs more and more tools for its own needs, and every single place needs to take care of picking the right manager, right `sudo`, right options... Meh).

happz replied:
Another point: are annotations necessary? Both mypy and pyright should be able to infer the type from the assignment alone.

carlosrodfern replied:
Yes, I will be using that utility. This is definitely a cleanup to do. I'll remove the type there as well.

lukaszachy replied:
@happz What seems better - Allow/use `PrepareInstall` to be run in almost any step or dig the 'dnf install' part from it into `tmt.utils` and use it from there?

For unrelated dist-git thingy I'll need to install few things as well :D

happz replied:
To me, it seemed better to generalize and convert `PrepareInstall` to use the generalized code.

Somewhere in my stash lies a draft of `tmt.package_managers` where I tried to cut things out. I wouldn't mind finishing it - e.g. the `avc` check needs `sestatus`, and I wasn't sure it's safe to rely on it being omnipresent. Plus we already install `rsync`, for example, so there is a need for "install me a package" primitive.

Then there is @thrix & his all-mighty install plugin effort, for which I was hoping to have this general library ready...

Now when the parallelization works are almost done, I guess it wouldn't be a bad thing to refresh this patch, there seems to be interest in it from multiple sides. And I still want to run full tmt test suite on my Ubuntu VM, soooo... `apt-get`, here we come :)

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1268,6 +1276,25 @@ def is_ready(self) -> bool:
+                self._ssh_guest(),
+                remote_commands.to_element()
+                ]
+            self._run_guest_command(ssh_command, silent=True)
```
I don't understand, is `guest.execute()` unusable for some reason?

carlosrodfern replied:
What I had in mind was that `execute` is a function the class makes available to other client code, and it is usually invoked after the guest is initialized. Basically, the expected sequence of states in a guest would be: `started`, `setup`, `ready for executes`. Therefore, since the expected use of `execute` is after initialization, I can see a developer coming later on and modifying `execute` with the assumption the guest is already fully setup, but in reality is not, so creating a bug in `setup` (`setup` calls `execute`, but `execute` expected `setup` already invoked).

So I went head and broke the dependency of `setup` in `execute`, and instead made it use the lower lever function `_run_guest_command`. My thought behind this is that the semantic of these functions and the expectations they create will drive the developers decision on where to put a change later on.

It is probably simpler to just invoke `execute`, though. It helps that it is just right there in the same class so bugs out of wrong assumptions are easy to spot.

carlosrodfern commented:
@happz @lukaszachy , overall, is this a good approach I should continue? setting up acl permissions globally at initialization time?

Other brainstroming ideas: setting up acl per plan workdir, or making sure the outputs file being generated are all with multi-user permissions from the start.

lukaszachy commented:
@carlosrodfern Thanks for the draft, but unfortunately I accumulated some other work I need to finish first. I'll take a better look next week.

IMO we don't need to be too specific here so setting it globally should be OK. I see no point in hardening /var/tmp/tmt on throwaway testing machine.

carlosrodfern commented:
@happz , I believe addressed all reviews. I also went thru the checklist, and added tests. If I missed something please let me know. Thank you for reviewing this PR.

happz commented:
> @happz , I believe addressed all reviews. I also went thru the checklist, and added tests. If I missed something please let me know. Thank you for reviewing this PR.

It seems to me another patch of yours sneaked into this PR, https://github.com/teemtee/tmt/pull/2659. It makes the review more complicated, but other than that...

@psss this PR was missing a milestone, would it still fit into 1.31 schedule?

carlosrodfern commented:
> > @happz , I believe addressed all reviews. I also went thru the checklist, and added tests. If I missed something please let me know. Thank you for reviewing this PR.
> 
> It seems to me another patch of yours sneaked into this PR, #2659. It makes the review more complicated, but other than that...
> 
> @psss this PR was missing a milestone, would it still fit into 1.31 schedule?

I separated the basic test for virtual provision into its own PR to get that one done sooner, and because it is indeed not directly related to this PR. That way the only test this PR would bring is the `umask` one. 
I also assumed you all needed time to look into this PR more carefully.

The PR that gets merged first will create a minor conflict for the other, but I'll fix it quickly for the second PR being merged.

happz commented:
> > > @happz , I believe addressed all reviews. I also went thru the checklist, and added tests. If I missed something please let me know. Thank you for reviewing this PR.
> > 
> > 
> > It seems to me another patch of yours sneaked into this PR, #2659. It makes the review more complicated, but other than that...
> > @psss this PR was missing a milestone, would it still fit into 1.31 schedule?
> 
> I separated the basic test for virtual provision into its own PR to get that one done sooner, and because it is indeed not directly related to this PR. That way the only test this PR would bring is the `umask` one. I also assumed you all needed time to look into this PR more carefully.
> 
> The PR that gets merged first will create a minor conflict for the other, but I'll fix it quickly for the second PR being merged.

I see. In that case, I would recommend merging https://github.com/teemtee/tmt/pull/2659 first because it contains just a single change, unlike this PR.

carlosrodfern commented:
> > > > @happz , I believe addressed all reviews. I also went thru the checklist, and added tests. If I missed something please let me know. Thank you for reviewing this PR.
> > > 
> > > 
> > > It seems to me another patch of yours sneaked into this PR, #2659. It makes the review more complicated, but other than that...
> > > @psss this PR was missing a milestone, would it still fit into 1.31 schedule?
> > 
> > 
> > I separated the basic test for virtual provision into its own PR to get that one done sooner, and because it is indeed not directly related to this PR. That way the only test this PR would bring is the `umask` one. I also assumed you all needed time to look into this PR more carefully.
> > The PR that gets merged first will create a minor conflict for the other, but I'll fix it quickly for the second PR being merged.
> 
> I see. In that case, I would recommend merging #2659 first because it contains just a single change, unlike this PR.

The pull request https://github.com/teemtee/tmt/pull/2659 has been merged. I rebased this PR, fixed some bug, setup the test based on the new tests, and pushed an updated commit.

psss commented:
Thanks for working on this. Let's try to squeeze this into 1.32.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -734,6 +735,13 @@ def start(self) -> None:
+        """
+        Setup the guest
+
+        Setup the guest after it has been started. It is called after start().
```
For better links with Sphinx, I'd suggest the following:

```
:py:meth:`Guest.start`
```

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1365,6 +1373,23 @@ def is_ready(self) -> bool:
+            return
+        if not self.facts.is_superuser and self.become:
+            assert self.facts.package_manager is not None
+            self.execute(
```
Please, add a TODO comment pointing to https://github.com/teemtee/tmt/pull/2557. Yet another place to switch to a true package manager.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1365,6 +1373,23 @@ def is_ready(self) -> bool:
+                    'acl'))
+            workdir_root = effective_workdir_root()
+            acl_setup_script = ShellScript(f'mkdir -p {workdir_root}')
+            acl_setup_script += ShellScript(f'setfacl -d -m o:rX {workdir_root}')
```
`ShellScript` supports multiline strings, i.e. the following might be more readable:

```python
self.execute(ShellScript(
    f"""
    mkdir -p {workdir_root};
    setfacl -d -m o:rX {workdir_root}
    """))
```

happz reviewed (approved):
I still don't like the fact plugin developer needs to call `setup()` explicitly, it's easy to forget. But we can fix that later (see https://github.com/teemtee/tmt/pull/2510#discussion_r1413551270).

happz commented:
> > I still don't like the fact plugin developer needs to call `setup()` explicitly, it's easy to forget. But we can fix that later (see [#2510 (comment)](https://github.com/teemtee/tmt/pull/2510#discussion_r1413551270)).
> 
> One idea that can help is to explicitly model the state machine this represents, with a variable or object representing the state, and then _action_ functions (e.g., `start`, `setup`, `wake`, `stop`, ...), that check preconditions like the current state, and transition it. That way you have code ensuring things are done in the right order of steps. Then you can add _helper_ functions (e.g., `get_ready`) that represent a "desired state" that help transition the state machine several states by calling the _action_ s functions in the right order.
> 
> Representing that state machine can help with readability for the developer, and correctness in general, by making the state explicit, and organizing the functions around that concept. Right now the state machine is implicitly represented in the values of the `Guest` properties. I think the current code is already very close.

I would go even further, with a typestate pattern where the state of an object is expressed by its type (e.g. https://cliffle.com/blog/rust-typestate/#a-simple-example-the-living-and-the-dead). TL;DR, "running" guest would be a different type than "not running". A running guest type would not even have methods like `start()`, "not running" guest would not have `stop()`, and so on. Once `wake()` or `setup()` get called, the proper type would be returned, something like `DeadGuest: start(self) -> RunningGuest`, `RunningGuest: def stop(self) -> DeadGuest`. Methods now accepting `Guest` would change their signatures to request the hypothetical `RunningAndReadyGuest` type only, plugins' `go()` method would return `RunningAndReadyGuest` instances instead of `None`, and `RunningGuest` would have `setup() -> RunningAndReadyGuest` method which would force a developer to call `Guest.start()` first to obtain `RunningGuest`, followed by call to its `setup()` to obtain `RunningAndReadyGuest` type that can be returned.

I already tried this out, it does not look that bad or complex, although I'm still far from a reviewable patch. But I liked the approach very much, it would detect violations in pre-commit time.

psss reviewed (approved):
Looks good. Thanks for fixing this!

psss commented:
/packit test
